File: QEasyWidgets/Components/Edit.py
from typing import Optional, overload
#from functools import singledispatchmethod
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *

from ..Common.Icon import *
from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from .Button import ClearButton, FileButton

##############################################################################################################################

class LineEditBase(QLineEdit):
    '''
    '''
    _clearButton = None
    _isClearButtonEnabled = False

    _fileButton = None
    _isFileButtonEnabled = False

    cursorPositionChanged = Signal(int, int)

    focusedIn = Signal()
    focusedOut = Signal()

    interacted = Signal()

    rectChanged = Signal(QRect)

    @singledispatchmethod
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)

        self.installEventFilter(self)
        self.textChanged.connect(lambda: self.interacted.emit())
        self.textChanged.connect(lambda: self.setClearButtonEnabled(True if len(self.text()) > 0 else False))

        self.spacer = QSpacerItem(0, 0, QSizePolicy.Expanding, QSizePolicy.Minimum)

        HBoxLayout = QHBoxLayout(self)
        HBoxLayout.setSpacing(0)
        HBoxLayout.setContentsMargins(0, 0, 0, 0)
        HBoxLayout.addSpacerItem(self.spacer)

        self.ToolTip = QToolTip() # TODO Change it to a custom tooltip

        self.IsAlerted = False

        StyleSheetBase.Edit.Apply(self)

    @__init__.register
    def _(self, arg__1: str, parent: Optional[QWidget] = None) -> None:
        self.__init__(parent)
        self.setText(arg__1)

    def showToolTip(self, content: Optional[str] = None) -> None:
        XPos = 0
        YPos = 0 - self.height()
        self.ToolTip.showText(self.mapToGlobal(QPoint(XPos, YPos)), content) if not self.ToolTip.isVisible() and content is not None else None

    def hideToolTip(self) -> None:
        self.ToolTip.hideText() if self.ToolTip.isVisible() else None

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        position = event.position()
        self.cursorPositionChanged.emit(position.x(), position.y())
        self.interacted.emit()
        super().mouseMoveEvent(event)

    def moveEvent(self, event: QMoveEvent) -> None:
        self.rectChanged.emit(self.rect())

    def resizeEvent(self, event: QResizeEvent) -> None:
        self.rectChanged.emit(self.rect())

    def dragEnterEvent(self, event: QDragEnterEvent):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()

    def dropEvent(self, event: QDropEvent):
        if event.mimeData().hasUrls():
            paths = [url.toLocalFile() for url in event.mimeData().urls()]
            self.setText(paths[0] if len(paths) == 1 else ", ".join(paths))

    def eventFilter(self, watched: QObject, event: QEvent) -> bool:
        if watched is self:
            if event.type() in (QEvent.DragEnter, QEvent.Drop):
                mime_data = event.mimeData()
                if mime_data.hasUrls():
                    event.acceptProposedAction()
        return super().eventFilter(watched, event)

    @property
    def clearButton(self):
        if self._clearButton is None:
            self.clearButton = ClearButton(self)
        return self._clearButton

    @clearButton.setter
    def clearButton(self, clearButton: ClearButton):
        clearButton.setBorderless(True)
        clearButton.setTransparent(True)
        clearButton.clicked.connect(self.clear)
        clearButton.clicked.connect(self.interacted.emit)
        self.layout().insertWidget(self.layout().indexOf(self.spacer) + 1, clearButton, alignment = Qt.AlignRight)
        self.setTextMargins(0, 0, clearButton.minimumSizeHint().width() + self.textMargins().right() + 3, 0)
        self._clearButton = clearButton

    def setClearButtonEnabled(self, enable: bool) -> None:
        self._isClearButtonEnabled = enable
        self.clearButton.setVisible(True if enable and self.hasFocus() else False)

    def isClearButtonEnabled(self) -> bool:
        return self._isClearButtonEnabled

    def focusInEvent(self, arg__1: QFocusEvent) -> None:
        self.focusedIn.emit()
        self.clearButton.show() if self.isClearButtonEnabled() else None
        super().focusInEvent(arg__1)

    def focusOutEvent(self, arg__1: QFocusEvent) -> None:
        self.focusedOut.emit()
        #self.clearButton.hide() if self.isClearButtonEnabled() else None
        super().focusOutEvent(arg__1)

    @property
    def fileButton(self):
        if self._fileButton is None:
            self.fileButton = FileButton(self)
        return self._fileButton

    @fileButton.setter
    def fileButton(self, fileButton: FileButton):
        fileButton.setBorderless(True)
        fileButton.setTransparent(True)
        fileButton.clicked.connect(self.interacted.emit)
        self.layout().insertWidget(self.layout().count(), fileButton, alignment = Qt.AlignRight)
        self.setTextMargins(0, 0, fileButton.minimumSizeHint().width() + self.textMargins().right() + 3, 0)
        self._fileButton = fileButton

    def setFileButtonEnabled(self, enable: bool) -> None:
        self._isFileButtonEnabled = enable
        self.fileButton.setVisible(True if enable else False)

    def isFileButtonEnabled(self) -> bool:
        return self._isFileButtonEnabled

    def setFileDialog(self, mode: str, fileType: Optional[str] = None, directory: Optional[str] = None, buttonTooltip: str = "Browse"):
        self.fileButton.setFileDialog(self, mode, fileType, directory, buttonTooltip)

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def setTransparent(self, transparent: bool) -> None:
        self.setProperty("isTransparent", transparent)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.Edit.Deregistrate(self)

    def setStyleSheet(self, styleSheet: str) -> None:
        AlertStyle = 'LineEditBase {border-color: red;}' if self.IsAlerted else ''
        super().setStyleSheet(styleSheet + AlertStyle)

    def alert(self, enable: bool, content: Optional[str] = None) -> None:
        self.IsAlerted = enable
        #StyleSheetBase.Edit.Apply(self)
        self.showToolTip(content) if enable else self.hideToolTip()

##############################################################################################################################

class TextEditBase(QTextEdit):
    '''
    '''
    keyEnterPressed = Signal()

    keyEnterBlocked = False

    @singledispatchmethod
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)

        #self.installEventFilter(self)

        HBoxLayout = QHBoxLayout(self)
        HBoxLayout.setSpacing(0)
        HBoxLayout.setContentsMargins(0, 0, 0, 0)

        StyleSheetBase.Edit.Apply(self)

    @__init__.register
    def _(self, text: str, parent: Optional[QWidget] = None) -> None:
        self.__init__(parent)
        self.setText(text)

    def keyPressEvent(self, e: QKeyEvent) -> None:
        if e.key() in (Qt.Key_Enter, Qt.Key_Return) and not (e.modifiers() == Qt.ShiftModifier) and not self.keyEnterBlocked:
            self.keyEnterPressed.emit()
        super().keyPressEvent(e)

    def blockKeyEnter(self, block: bool) -> None:
        self.keyEnterBlocked = block

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def setTransparent(self, transparent: bool) -> None:
        self.setProperty("isTransparent", transparent)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.Edit.Deregistrate(self)

##############################################################################################################################

File: c++/QEasyWidgets/Components/Edit.cpp
#include "Edit.h"

#include <QMimeData>
#include <QMouseEvent>
#include <QKeyEvent>

#include "../Common/StyleSheet.h"


/**
 * LineEditBase implementation
 */

LineEditBase::LineEditBase(QWidget *parent)
    : QLineEdit(parent)
    , m_clearButton(nullptr)
    , m_fileButton(nullptr)
    , m_spacer(nullptr)
    , m_layout(nullptr)
    , m_toolTip(nullptr)
    , m_isClearButtonEnabled(false)
    , m_isFileButtonEnabled(false)
    , m_isAlerted(false) {
    init();
}

LineEditBase::LineEditBase(const QString &text, QWidget *parent)
    : LineEditBase(parent) {
    setText(text);
}

void LineEditBase::setClearButtonEnabled(bool enable) {
    m_isClearButtonEnabled = enable;
    if (m_clearButton) {
        m_clearButton->setVisible(enable && hasFocus());
    }
}

bool LineEditBase::isClearButtonEnabled() const {
    return m_isClearButtonEnabled;
}

void LineEditBase::setFileButtonEnabled(bool enable) {
    m_isFileButtonEnabled = enable;
    if (m_fileButton) {
        m_fileButton->setVisible(enable);
    }
}

bool LineEditBase::isFileButtonEnabled() const {
    return m_isFileButtonEnabled;
}

void LineEditBase::setFileDialog(QFileDialog::FileMode mode, const QString &fileType, const QString &directory, const QString &buttonTooltip) {
    if (!m_fileButton) {
        // Create file button if it doesn't exist
        m_fileButton = new FileButton(this);
        m_fileButton->setBorderless(true);
        m_fileButton->setTransparent(true);
        connect(m_fileButton, &QPushButton::clicked, this, [this]() {
            emit interacted();
        });
        m_layout->addWidget(m_fileButton, 0, Qt::AlignRight);
    }

    m_fileButton->setFileDialog(this, mode, fileType, directory, buttonTooltip);
    setFileButtonEnabled(true);
}

void LineEditBase::setBorderless(bool borderless) {
    setProperty("isBorderless", borderless);
    style()->unpolish(this);
    style()->polish(this);
}

void LineEditBase::setTransparent(bool transparent) {
    setProperty("isTransparent", transparent);
    style()->unpolish(this);
    style()->polish(this);
}

void LineEditBase::clearDefaultStyleSheet() {
    StyleSheetBase::deregistrate(this);
}

void LineEditBase::alert(bool enable, const QString &content) {
    m_isAlerted = enable;
    if (enable && !content.isEmpty()) {
        m_toolTip->showText(mapToGlobal(QPoint(0, -height())), content);
    } else {
        m_toolTip->hideText();
    }
    // Force stylesheet update
    setStyleSheet(styleSheet());
}

void LineEditBase::mouseMoveEvent(QMouseEvent *event) {
    QLineEdit::mouseMoveEvent(event);
    emit cursorPositionChanged(event->position().x(), event->position().y());
    emit interacted();
}

void LineEditBase::moveEvent(QMoveEvent *event) {
    QLineEdit::moveEvent(event);
    emit rectChanged(rect());
}

void LineEditBase::resizeEvent(QResizeEvent *event) {
    QLineEdit::resizeEvent(event);
    emit rectChanged(rect());
}

void LineEditBase::focusInEvent(QFocusEvent *event) {
    QLineEdit::focusInEvent(event);
    emit focusedIn();
    if (m_clearButton && m_isClearButtonEnabled) {
        m_clearButton->show();
    }
}

void LineEditBase::focusOutEvent(QFocusEvent *event) {
    QLineEdit::focusOutEvent(event);
    emit focusedOut();
    if (m_clearButton && m_isClearButtonEnabled) {
        m_clearButton->hide();
    }
}

void LineEditBase::dragEnterEvent(QDragEnterEvent *event) {
    if (event->mimeData()->hasUrls()) {
        event->acceptProposedAction();
    }
    QLineEdit::dragEnterEvent(event);
}

void LineEditBase::dropEvent(QDropEvent *event) {
    if (event->mimeData()->hasUrls()) {
        QList<QUrl> urls = event->mimeData()->urls();
        if (!urls.isEmpty()) {
            QStringList paths;
            for (const QUrl &url : urls) {
                paths.append(url.toLocalFile());
            }
            setText(paths.join(", "));
        }
    }
    QLineEdit::dropEvent(event);
}

void LineEditBase::init() {
    installEventFilter(this);

    connect(this, &QLineEdit::textChanged, this, [this]() {
        emit interacted();
        setClearButtonEnabled(!text().isEmpty());
    });

    m_toolTip = new ToolTipBase(this);
    installEventFilter(new ToolTipEventFilter(this, m_toolTip));

    m_layout = new QHBoxLayout(this);
    m_layout->setSpacing(0);
    m_layout->setContentsMargins(0, 0, 0, 0);

    m_spacer = new QSpacerItem(0, 0, QSizePolicy::Expanding, QSizePolicy::Minimum);
    m_layout->addItem(m_spacer);

    StyleSheetBase::apply(this, StyleSheetBase::Edit);
}

QSize LineEditBase::sizeHint() const {
    QSize hint = QLineEdit::sizeHint();
    if (width() > 0) {
        hint.setWidth(qMax(hint.width(), width()));
    }
    return hint;
}

/**
 * TextEditBase implementation
 */

TextEditBase::TextEditBase(QWidget *parent)
    : QTextEdit(parent)
    , m_keyEnterBlocked(false) {
    init();
}

TextEditBase::TextEditBase(const QString &text, QWidget *parent)
    : TextEditBase(parent) {
    setText(text);
}

void TextEditBase::blockKeyEnter(bool block) {
    m_keyEnterBlocked = block;
}

bool TextEditBase::isKeyEnterBlocked() const {
    return m_keyEnterBlocked;
}

void TextEditBase::setBorderless(bool borderless) {
    setProperty("isBorderless", borderless);
    style()->unpolish(this);
    style()->polish(this);
}

void TextEditBase::setTransparent(bool transparent) {
    setProperty("isTransparent", transparent);
    style()->unpolish(this);
    style()->polish(this);
}

void TextEditBase::clearDefaultStyleSheet() {
    StyleSheetBase::deregistrate(this);
}

void TextEditBase::keyPressEvent(QKeyEvent *event) {
    QTextEdit::keyPressEvent(event);
    if ((event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return) &&
        !(event->modifiers() & Qt::ShiftModifier) && !m_keyEnterBlocked) {
        emit keyEnterPressed();
    }
}

void TextEditBase::init() {
    QHBoxLayout *layout = new QHBoxLayout(this);
    layout->setSpacing(0);
    layout->setContentsMargins(0, 0, 0, 0);

    StyleSheetBase::apply(this, StyleSheetBase::Edit);
}